Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -0,0 +1,276 @@
+ * |  |                   +-----+                 |   |                   +-----+                 | |
+ * |  +-------------------------------------------+   +-------------------------------------------+ |
+ * +------------------------------------------------------------------------------------------------+
+ * ¹ High speed channels are only available when SOC_LEDC_SUPPORT_HS_MODE is defined as 1
+ * ² The ESP32C3 does only support six channels, so 6 and 7 are not available on that SoC
+ * 
+ * The nomenclature of timers in the high speed / low speed blocks is a bit misleading as the idf api 
+ * speaks of "speed mode", which, to me, implies that this would be a mode configurable in a specific timer
+ * while in reality, it does select a block of timers.
+ * 
+ * Maximum Timer width for PWM:
+ * ============================
+ * esp32   SOC_LEDC_TIMER_BIT_WIDE_NUM  (20)
+ * esp32c3 SOC_LEDC_TIMER_BIT_WIDE_NUM  (14)
+ * esp32s2 SOC_LEDC_TIMER_BIT_WIDE_NUM  (14)
+ * esp32s3 SOC_LEDC_TIMER_BIT_WIDE_NUM  (14)
+ * 
+ * Number of Channels:
+ * ===================
+ * esp32   SOC_LEDC_CHANNEL_NUM         (8)
+ * esp32c3 SOC_LEDC_CHANNEL_NUM         (6)
+ * esp32s2 SOC_LEDC_CHANNEL_NUM         (8)
+ * esp32s3 SOC_LEDC_CHANNEL_NUM 		(8)
+ *
+ * Some SoSs support a mode called HIGHSPEED_MODE which is essentially another full block of PWM hardware 
+ * that adds SOC_LEDC_CHANNEL_NUM channels. 
+ * Those Architectures have SOC_LEDC_SUPPORT_HS_MODE defined as 1.
+ * In esp-idf-4.3 that's currently only the esp32 SOC 
+ * 
+ * Supports highspeed mode:
+ * ========================
+ * esp32 SOC_LEDC_SUPPORT_HS_MODE	(1)
+ * 
+ * hardware technical reference: 
+ * =============================
+ * https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#ledpwm
+ * 
+ * Overview of the whole ledc-system here: 
+ * https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/ledc.html
+ * 
+ * the ledc interface also exposes some advanced functions such as fades that are then done in hardware. 
+ * ToDo: implement a Sming interface for fades
+ * 
+ */
+
+#include <HardwarePWM.h>
+#include <debug_progmem.h>
+#include <driver/periph_ctrl.h>
+#include <driver/ledc.h>
+#include <esp_err.h>
+#include <hal/ledc_types.h>
+
+#define DEFAULT_RESOLUTION ledc_timer_bit_t(10)
+#define DEFAULT_PERIOD 200
+namespace
+{
+ledc_channel_t pinToChannel(uint8_t pin)
+{
+	return ledc_channel_t(pin % 8);
+}
+
+ledc_mode_t pinToGroup(uint8_t pin)
+{
+	return ledc_mode_t(pin / 8);
+}
+
+ledc_timer_t pinToTimer(uint8_t pin)
+{
+	return ledc_timer_t((pin / 2) % 4);
+}
+
+uint32_t periodToFrequency(uint32_t period)
+{
+	return (period == 0) ? -1 : (1000000 / period);
+}
+
+uint32_t frequencyToPeriod(uint32_t freq)
+{
+	return (freq == 0) ? -1 : (1000000 / freq);
+}
+
+uint32_t maxDuty(ledc_timer_bit_t bits)
+{
+	return (1U << bits) - 1;
```

Comment:
Please, check the return value to be valid uint32_t.

Reply:
hmm.. I can't see how it would not (I know, famous last words) - the only error condition here would be bits>32 - so I would tend to check for bits <= DEFAULT_RESOLUTION (which is currently set to 10).